File: web/test/measurement.test.ts
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { test } from "node:test";
import { fileURLToPath } from "node:url";

import {
  addSecondReading,
  captureFirstReading,
  confirmMeasurement,
  roundHalfUpMean,
} from "../src/domain/measurement.ts";

test("mean follows the shared ROUND_HALF_UP contract", () => {
  for (const [first, second, expected] of contractCases()) {
    assert.equal(roundHalfUpMean(first, second), expected, `${first}, ${second}`);
  }
});

test("calculates all fields and only confirmation creates persistible type", () => {
  const first = captureFirstReading({ systolic: 121, diastolic: 79, pulse: 61 });
  const calculated = addSecondReading(first, { systolic: 122, diastolic: 82, pulse: 62 });

  assert.deepEqual(calculated.result, { systolic: 122, diastolic: 81, pulse: 62 });
  assert.deepEqual(confirmMeasurement(calculated), {
    kind: "confirmed-measurement",
    values: { systolic: 122, diastolic: 81, pulse: 62 },
  });
});

test("confirmed measurement does not retain original readings", () => {
  const confirmed = confirmMeasurement(
    addSecondReading(
      captureFirstReading({ systolic: 120, diastolic: 80, pulse: 60 }),
      { systolic: 122, diastolic: 82, pulse: 62 },
    ),
  );

  assert.deepEqual(Object.keys(confirmed).sort(), ["kind", "values"]);
});

test("validates only documented unequivocal invariants", () => {
  assert.throws(() => captureFirstReading({ systolic: 0, diastolic: 80, pulse: 60 }));
  assert.throws(() => captureFirstReading({ systolic: 120, diastolic: 0, pulse: 60 }));
  assert.throws(() => captureFirstReading({ systolic: 120, diastolic: 80, pulse: 0 }));
  assert.throws(() => captureFirstReading({ systolic: 80, diastolic: 80, pulse: 60 }));
  assert.throws(() => captureFirstReading({ systolic: 120.5, diastolic: 80, pulse: 60 }));
});

function contractCases(): Array<[number, number, number]> {
  const contractUrl = new URL("../../contracts/measurement-mean-cases.csv", import.meta.url);
  return readFileSync(fileURLToPath(contractUrl), "utf8")
    .trim()
    .split("\n")
    .slice(1)
    .map((line) => line.split(",").map(Number) as [number, number, number]);
}

File: android/domain/src/test/kotlin/com/mitension/domain/MeasurementTest.kt
package com.mitension.domain

import java.nio.file.Files
import java.nio.file.Path
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

class MeasurementTest {
    @Test
    fun `mean follows the shared ROUND_HALF_UP contract`() {
        contractCases().forEach { (first, second, expected) ->
            assertEquals(expected, roundHalfUpMean(first, second), "$first, $second")
        }
    }

    @Test
    fun `calculates all fields and only confirmation creates persistible type`() {
        val firstStep = FirstReadingCaptured(Reading(121, 79, 61))
        val calculated = firstStep.addSecondReading(Reading(122, 82, 62))

        assertEquals(MeasurementValues(122, 81, 62), calculated.result)
        assertEquals(ConfirmedMeasurement(MeasurementValues(122, 81, 62)), calculated.confirm())
    }

    @Test
    fun `confirmed measurement does not retain original readings`() {
        val confirmed = FirstReadingCaptured(Reading(120, 80, 60))
            .addSecondReading(Reading(122, 82, 62))
            .confirm()

        assertEquals(MeasurementValues(121, 81, 61), confirmed.values)
    }

    @Test
    fun `validates only documented unequivocal invariants`() {
        assertFailsWith<IllegalArgumentException> { Reading(0, 80, 60) }
        assertFailsWith<IllegalArgumentException> { Reading(120, 0, 60) }
        assertFailsWith<IllegalArgumentException> { Reading(120, 80, 0) }
        assertFailsWith<IllegalArgumentException> { Reading(80, 80, 60) }
    }

    private fun contractCases(): List<Triple<Int, Int, Int>> {
        val path = Path.of(requireNotNull(System.getProperty("measurement.contract.path")))
        return Files.readAllLines(path)
            .drop(1)
            .filter(String::isNotBlank)
            .map { line ->
                val (first, second, expected) = line.split(',').map(String::toInt)
                Triple(first, second, expected)
            }
    }
}
